### PR TITLE
Max score

### PR DESCRIPTION
**Pull request by tsivert**
This is a portion of the logic described here: https://h5ptechnology.atlassian.net/browse/HFP-1806 and https://h5ptechnology.atlassian.net/browse/HFP-1807.
The rest is in the pull request for h5p/h5p-php-library.

Added support for processing presave logic related to max score.

### CODE REVIEW

**Comment by falcon-git** [comment]
Review issue:
https://h5ptechnology.atlassian.net/browse/HFP-1992

File: scripts/h5peditor-pre-save.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  };
+
+  /**
+   * Check if score is valid
+   *
+   * @static
+   * @param score
+   * @returns {boolean}
+   */
+  Presave.isScoreValid = function (score) {
```
Comment (thomasmars) [inline scripts/h5peditor-pre-save.js]: This function is never used. Is this intended ?
 Exception is thrown if max score is 0 ref: https://h5ptechnology.atlassian.net/browse/HFP-1807. Not sure if this should be an acceptance criteria though, as there are many times where a content type would return 0 maxscore, e.g. CP without any scorable interactions.

**Comment by tsivert** [comment]
It's kinda important to know if a score of 0 should automatically throw an exception or not. As you mention, a framework like CP, can give a score of 0 if the frameworks used in the CP don't give a score. Then what? Should this perhaps be framework specific and handled in the frameworks where this is relevant? 
What do you think @falcon-git @thomasmars ?

**Comment by falcon-git** [comment]
I think the order of which we execute various operators in the English language differs. To compensate for this I've added parenthesis around the part that should be executed first. Hopefully we'll now get the same result for the acceptance criteria in question.

File: scripts/h5peditor-pre-save.js (hunk 1 above)
Comment (tsivert) [reply scripts/h5peditor-pre-save.js]: As I remember I used the method in an early version of the code and left it there for convenience purposes. But I can remove it as it's not currently being used.

**Comment by tsivert** [comment]
Sorry, I misunderstood your question for a example implementation. I thought you meant an example where the score was being calculated. 

I don't have an example implementation for you. I've been developing this in our company platform which is not publicly available and on a local installation of Drupal. But my knowledge of Drupal is at a minimum so I haven't made any changes to the logic of persisting and retrieving the max score. It only sends the new parameter to the server but does nothing with it.

Of course I'll add the docs for the functions you mentioned. But can you please answer the question I asked on the pull request for h5p-php-library(https://github.com/h5p/h5p-php-library/pull/50). The answer there determines if I need to move the new functions or keep them in this repo.

**Comment by thomasmars** [comment]
I see, thanks. I've answered the question, and I'll make an example implementation myself and try this out before I approve it when I have time.

**Comment by tsivert** [comment]
Just to keep the ball rolling and to avoid misunderstandings. What exactly are you expecting in the example implementation? Is there any way I can contribute to this?

**Comment by tsivert** [comment]
I'm a noob to this way of working(writing code for other peoples repos) so I just have to ask: am I fixing the conflicts or are you doing it? :)

**Comment by thomasmars** [comment]
An example implementation would typically have to add the max score hidden input element to the H5P form and make sure that it is passed to the init() function of the editor. In Drupal 7 this means changing the form and the application.js script for initializing the editor. Furthermore checking that max score is retrieved, calculated and passed appropriately to the save content endpoint, and that this will work for compound content types. The Drupal 7 branches are available at https://cgit.drupalcode.org/h5p if you want to contribute to this.
Typically you would fix any conflicts, but if they appear right before the merge the maintainer can resolve them. Thanks 👍

File: h5peditor.class.php [hunk 2]
```diff
@@ -640,4 +646,49 @@ public function mergeLocalLibsIntoCachedLibs($local_libraries, &$cached_librarie
...
+      }
+    }
+    return $this->h5p->fs->hasPresave($libraryName);
+  }
+
+  /**
+   * Adds the path to the presave.js file to the list of dependency assets for the library
+   *
+   * @param array $assets
+   * @param string $library
```
Comment (thomasmars) [inline h5peditor.class.php]: $library says it is a string, but actually expects an array to be passed in.

File: h5peditor.class.php [hunk 3]
```diff
@@ -353,6 +354,11 @@ public function getLibraryData($machineName, $majorVersion, $minorVersion, $lang
 
     // Get list of JS and CSS files that belongs to the dependencies
     $files = $this->h5p->getDependenciesFiles($libraries, $prefix);
+    $libraryName = H5PCore::libraryToString(compact('machineName', 'majorVersion', 'minorVersion'), true);
+    if( $this->hasPresave($libraryName) === true ){
+      $library = $this->h5p->loadLibrary($machineName, $majorVersion, $minorVersion);
```
Comment (thomasmars) [inline h5peditor.class.php]: Loading library will cause several queries to the database, is this really necessary to add the presave file ?

**Comment by tsivert** [comment]
Hi.
I'll fix the misleading PhpDoc and change the querying against the database right away. 

But the crucial part I can't reproduce. I've create a CP with 4 different content types. Then I go into edit mode where I see, according to my network tab in Chrome, that the presave file for each content type is loaded even before I start editing anything. I then edit only on the first slide (adding one more alternative that will give a score) and click save. The new max score is now calculated to be one more point than the previous one. 
Do you have a example H5P that I can use to reproduce?

File: h5peditor.class.php (hunk 3 above)
Comment (tsivert) [reply h5peditor.class.php]: The reason for the query is to get the patchVersion which is used in the "version" attribute when creating the path to the dependent files for a library(h5p.classes.php, line 2117). This is logic that all the other libraries use and I just "tagged" along it. 
The alternatives would be to either read the library.json file from disk or skip the patchVersion in the url to the file.

**Comment by thomasmars** [comment]
You're right, it is always loaded eventually. However it is loaded as the last script, so if you save the Course Presentation before all scripts have loaded you could get an inaccurate max score since some of the presave.js scripts might not have loaded. We should probably make sure that all presave.js scripts are loaded before you can save the content. I think this is out of the scope for this pull request though, you can have a look at it if you want. I'll try to get this merged in today, good work :)

**Comment by tsivert** [comment]
Nice! And thank you. :)

With this logic in place I'll create pull requests for all the libraries that calculates a score.

File: h5peditor.class.php (hunk 3 above)
Comment (thomasmars) [reply h5peditor.class.php]: The library info for the main library is already loaded as part of the findEditorLibraries() in $libraries, so we're fetching the data an additional time. It seems like we're loading the libraries for all the editor libraries though, so not doing it here is only a small performance improvement, there are probably other things that would give us better gains, so we can skip this.

Comment (tsivert) [reply h5peditor.class.php]: Ahhh. I oversaw that. I'll take a look at it at a later stage and fix it.

**Pull request merged**